File: lib/ast/lazy_builder.rb
module AST
  class LazyBuilder
    def build_keyword(name)
      -> builder { builder.build_keyword(name) }
    end

    def build_sequence(expressions)
      -> builder { builder.build_sequence(expressions.map { |e| e.call(builder) }) }
    end

    def build_variable(name)
      -> builder { builder.build_variable(name) }
    end
  end
end

File: src/lazy_builder.js
export default class {
  buildKeyword(name) {
    return builder => builder.buildKeyword(name);
  }

  buildSequence(expressions) {
    return builder => builder.buildSequence(expressions.map(e => e(builder)));
  }

  buildVariable(name) {
    return builder => builder.buildVariable(name);
  }
}
